### PR TITLE
Test Solution for Forem-link Embed issues on Canary2

### PR DESCRIPTION
**Pull request by msarit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [X] Testing a Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Several members of Fleet Forems have complained about not being able to embed valid, accurate forem-specific links (e.g. post, comment or profile links). This PR tests a potential solution for this problem JUST on the Canary2 forem.

## QA Instructions, Screenshots, Recordings
NONE. The QA shall be done on https://canary2.forem.wtf after this PR is deployed to that forem.

### UI accessibility concerns?
None; all backend changes.

## Added/updated tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?
If the testing is successful, I will discuss with the relevant team members
about rolling this solution out to the rest of the fleet.

### CODE REVIEW

File: app/lib/url.rb
Comment (djuber) [inline app/lib/url.rb:8]: Why active record connected, and not app initialized as before? 

Asset precompilation (in docker build/buildkite, and in travis) initializes the app, so we'll call URL.url (from the liquid tag initializer) whether we have a db or not.

Since we want "live" settings from the db in production when initializing the liquid tags, we can't ask, in an initializer, if we're initialized (that answer is always false), but don't want to add a database in docker just to support assets precompilation in the builder. 

So here, when we don't have a database to check, fallback to the environment, otherwise, whenever possible, load the domain set in the settings table (which has a built in fallback to the env var), and use that value.

File: spec/lib/url_spec.rb
Comment (djuber) [inline spec/lib/url_spec.rb:7]: the purpose of this (never called) APP_DOMAIN config value is to contrast that with the app_domain setting in the test for URL.domain, "returns the value of Settings::General"

**Review by dscottS3 — APPROVED** [review]

**Review by djuber — APPROVED** [review]
Approving with the understanding that the intent is to deploy to canary to confirm this both addresses the original issue with embeds, and doesn't break anything else, before merging to main/deploying generally.

**Review by Zhao-Andy — APPROVED** [review]
Looks to be working on Canary!

**Comment by msarit** [comment]
@jeremyf @dan @jgaskins @dscottS3 Hello people!

I surfaced the following questions in Slack, but then thought it better to have the discussion here.

Re: this PR. My changes have been tested on Canary2 and FINALLY the forem-specific embeds are working! 🎉 

Moving ahead, what’s our approach for deploying this PR to the fleet? Do you think the PR been vetted enough - bearing in mind that its purpose was to test a solution? What I'm really asking is: is the solution robust enough? @djuber provided me a lot of context and support, but I still want to leave room for a possible "better" solution.

Also, do we also want to deploy this PR to DEV specifically (which is working fine without it)?

**Review by jeremyf — APPROVED** [review]

**Comment by jeremyf** [comment]
Way to go @msarit on spearheading this effort!

**Comment by jeremyf** [comment]
I defer to @djuber and @Zhao-Andy and @jgaskins in regards to deploy strategy.  I'm happy to help test, but this is outside my realm of understanding (happy to learn more)

**Comment by dscottS3** [comment]
> @jeremyf @dan @jgaskins @dscottS3 Hello people!
> 
> I surfaced the following questions in Slack, but then thought it better to have the discussion here.
> 
> Re: this PR. My changes have been tested on Canary2 and FINALLY the forem-specific embeds are working! 🎉
> 
> Moving ahead, what’s our approach for deploying this PR to the fleet? Do you think the PR been vetted enough - bearing in mind that its purpose was to test a solution? What I'm really asking is: is the solution robust enough? @djuber provided me a lot of context and support, but I still want to leave room for a possible "better" solution.
> 
> Also, do we also want to deploy this PR to DEV specifically (which is working fine without it)?

Maybe we can deploy this to [community.benhalpern.com](http://community.benhalpern.com/) and do a bit more testing there. I'd say once it's validated there we'd be in good shape.

**Comment by jgaskins** [comment]
> Moving ahead, what’s our approach for deploying this PR to the fleet? Do you think the PR been vetted enough - bearing in mind that its purpose was to test a solution? What I'm really asking is: is the solution robust enough? @djuber provided me a lot of context and support, but I still want to leave room for a possible "better" solution.

Just a few questions because I feel like I'm missing a lot of this story due to it happening in multiple places.

@msarit @djuber @Zhao-Andy Can we outline the steps here that were taken to test this on Canary2 and what specific problem we were solving with it? Arit already covered the high-level impact of the problem, but understanding how we arrived at this solution would be great.

I see that we confirm that it works on Canary2 with this change, but I don't see confirmation that this functionality did _not_ work on Canary2 when deployed from `main` — is that also somewhere?

I'm just trying to make sure we have a consolidated record of what the definition of "success" is here. Some of this seems to have been discussed on Slack, but folks who have access to this thread may not have access to that Slack conversation for future debugging and, for those who do, a link to that discussion would be really helpful.

**Comment by msarit** [comment]
@jgaskins

Problem: In Forem Fleet (not DEV), when members created embeds of their forem's links (such as posts, comments, etc), they received a false "Embed URL not valid" error. In other words, the regexp that picked up forem-specific embeds (`%r{#{URL.url}/?}`) was rejecting valid forem urls.

Problem-Solving: We (myself, Dwight and AndyZ) discovered that `URL.url` was most-likely NOT returning the forem's domain. We needed Andy to verify this on the fleet forems.

Solution: With Dan's help, we produced a more reliable and consistent way to return the forem's actual domain (and not a fallback) - this PR. Then we deployed this PR to Canary2 for testing, and now embedding is working on Canary2 (wasn't before).

**Comment by jgaskins** [comment]
@msarit This is perfect! Thanks!

**Comment by djuber** [comment]
testing on benhalpern community is just validating that deploying this change to heroku doesn't fail? 

As I understand the problem, embeds should work there _before_ we deploy and that's a smoke test to confirm we're not breaking anything

**Comment by Zhao-Andy** [comment]
> testing on benhalpern community is just validating that deploying this change to heroku doesn't fail?
> 
> As I understand the problem, embeds should work there _before_ we deploy (already, now) and that's a smoke test to confirm we're not breaking anything

Yep exactly, just a test to see if it doesn't break anything on Heroku. It's deployed on benhalpern community now and the site seems to work fine. Embed also continues to work. I think we're good to go @msarit 👍

**Comment by msarit** [comment]
@Zhao-Andy Thank you!

**Comment by michael-tharrington** [comment]
Just to note that I've hit up the user who asked about this issue here - https://community.codenewbie.org/michaeltharrington/comment/ab2

**Pull request merged**
